### PR TITLE
[2.19.x] DDF-UI-248 Update UI automatically on share/unshare/delete

### PR DESCRIPTION
**Pull request by cassandrabailey293**
backport of https://github.com/codice/ddf-ui/pull/295

#### What does this PR do?
Add Server-Sent Event messaging to DDF, and uses it to automatically update search forms, workspaces, and result forms on share, unshare, and delete operations 

#### Who is reviewing it? 

@andrewzimmer @bdeining @hayleynorton @zta6 @abel-connexta 

#### Select relevant component teams: 

@codice/ui 

#### Ask 2 committers to review/merge the PR and tag them here.

@pklinef 
@rymach 

### Setup
* build 
* before starting ddf go into /etc/ directory, update users.attributes file to include
```
"test" : {
       "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress":"test@localhost.local",
       "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/role": ["manager","viewer","systembundles"]
   }
```
and in users.properties files add
`test=test,group,manager,viewer,systembundles,ssh`

* start ddf, profile install, and start the dev server
* log into the test user on `localhost:8993/search/catalog`
* open another window and go to `localhost:8080/search/catalog`

### Share/Unshare
0. Navigate to the workspaces collection page on `admin` user's instance
1. Create a new workspace on `test` user's instance
2. Share the workspace with the `admin` group (should be there already)
3. Verify on the `admin` instance that `test` user's workspace shows up automatically (takes 3 seconds)
4. Share the other way around (from `admin` to `test@localhost.local`) and verify it shows up on `test`'s workspace collection automatically

Repeat these steps for unshare. The workspace should disappear from the user to whom the workspace was unshared.

### Delete
0-3. Do steps 0-3 from *Share/Unshare*
4. Delete the workspace on `test` user's page
5. Verify on the `admin` instance that `test` user's workspace disappears automatically (takes 3 seconds)
6. Delete the other way around and verify

Repeat these steps for search forms and result forms.

### Workspace Search
0. Open a workspace on `admin`'s instance and Click on the Search DDF button
1. Create a new search form on `test` user's instance
2. Share the search form with the `admin` group (should be there already)
3. Verify on the `admin` instance that `test` user's search form shows up in the three-dot menu drop down when you click "Use Other Search Form" (takes 3 seconds)
4. Create a new result form on the `test` user's instance
5. Share that result form with `admin`
6. Verify in `admin` user's instance that the result form shows up in the Result Form dropdown in the advanced search from.  

#### What are the relevant tickets?
Fixes: https://github.com/codice/ddf-ui/issues/248

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+              // Sending SSE heartbeat
+              out.write(": \n\n");
+              if (out.checkError()) {
+                // Subscriber error, break out of loop
+                break;
+              }
+              Thread.sleep(1000);
+            }
+            listeners.remove(out);
+            return "";
```
Comment (cassandrabailey293) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java:47]: I'm not sure what I should be returning here (and on line 51). we're not using these return values at all. @bdeining any thoughts?

File: ui/packages/catalog-ui-search/src/main/webapp/js/EventSourceUtil.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+var EventSource = require('eventsource')
+
+const ORIGIN_HEADER = 'https://localhost:8993/'
```
Comment (cassandrabailey293) [inline ui/packages/catalog-ui-search/src/main/webapp/js/EventSourceUtil.js:3]: I know this shouldn't be hard-coded but I'm not sure how to get the right value from this. @bdeining any ideas?

Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/js/EventSourceUtil.js:3]: I think we need to use ` credentials: 'same-origin'` https://github.com/codice/ddf-ui/blob/master/ui-frontend/packages/catalog-ui-search/src/main/webapp/react-component/utils/fetch/fetch.tsx#L55

Comment (cassandrabailey293) [reply ui/packages/catalog-ui-search/src/main/webapp/js/EventSourceUtil.js:3]: i guess this `credentials` parameter isn't technically part of the headers - the [EventSource](https://github.com/EventSource/eventsource) object I think only takes a url and an object representing headers but I'll try it this way.

Comment (cassandrabailey293) [reply ui/packages/catalog-ui-search/src/main/webapp/js/EventSourceUtil.js:3]: looks like it worked 👍

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.collection.js [hunk 3]
```diff
@@ -49,6 +50,21 @@ module.exports = Backbone.AssociatedModel.extend({
       this.addResultForms()
       this.doneLoading()
     })
+    let self = this
```
Comment (andrewzimmer) [inline ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.collection.js:53]: ✏️ think you could also define the `onMessage` function here and pass it in to avoid this assignment

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form.collection.js [hunk 4]
```diff
@@ -66,6 +67,22 @@ module.exports = Backbone.AssociatedModel.extend({
       this.addAllForms()
       this.doneLoading()
     })
+
+    const onMessage = () => {
+      if (promiseIsResolved === true) {
```
Comment (rymach) [inline ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form.collection.js:72]: consider combining w/ duplicate code

**Comment by rymach** [comment]
how will this impact network perf? 
will this be incredibly chatty if many changes to workspaces/resultforms/etc are happening? 
all clients will now know when other users are changing workspaces/etc. it's a small detail, but it is something they wouldn't have known before

File: ui/packages/catalog-ui-search/src/main/webapp/js/EventSourceUtil.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  'X-Requested-With': REQUEST_HEADER,
+}
+
+var sources = []
+
+module.exports = {
+  createEventListener(type, handlers) {
+    const { onMessage } = handlers
+    if (sources.length != 0) {
+      sources[0].addEventListener(type, event => {
```
Comment (rymach) [inline ui/packages/catalog-ui-search/src/main/webapp/js/EventSourceUtil.js:14]: will sources only ever have one element?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+          return "Event Source configuration error: Server error";
+        });
+  }
+
+  public static void notifyListeners(String type) {
+    ExecutorService es = Executors.newSingleThreadExecutor();
+    es.submit(
+        () -> {
+          try {
+            Thread.sleep(3000);
```
Comment (rymach) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java:62]: why is this needed?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+          } catch (Exception e) {
+            LOGGER.error("Event Source configuration error");
+          }
+          res.status(500);
+          return "Event Source configuration error: Server error";
+        });
+  }
+
+  public static void notifyListeners(String type) {
+    ExecutorService es = Executors.newSingleThreadExecutor();
```
Comment (rymach) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java:58]: if perf is an issue consider a threadpool

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+import org.slf4j.LoggerFactory;
+import spark.servlet.SparkApplication;
+
+public class EventApplication implements SparkApplication {
+
+  private static final List<PrintWriter> listeners = new ArrayList<>();
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(EventApplication.class);
+
+  private static final String HEARTBEAT = ": \n\n";
```
Comment (rymach) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java:20]: consider adding a comment explaining why this particular literal

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+            while (true) {
+              out.write(HEARTBEAT);
+              if (out.checkError()) {
+                // Subscriber error, break out of loop
+                break;
+              }
+              Thread.sleep(1000);
+            }
+            listeners.remove(out);
+            return "Event Source setup successful";
```
Comment (rymach) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java:48]: is this message misleading?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            res.type("text/event-stream; charset=UTF-8");
+            res.header("Connection", "keep-alive");
+            res.header("Cache-Control", "no-cache");
+            res.status(200);
+            PrintWriter out = res.raw().getWriter();
+            synchronized (listeners) {
+              listeners.add(out);
+            }
+            out.write(HEARTBEAT);
+            out.flush();
```
Comment (rymach) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java:37]: are these needed if we write the HEARTBEAT in a loop below? is the flush needed below?

**Comment by rymach** [comment]
does this reconnect if the connection fails?

**Comment by rymach** [comment]
https://github.com/EventSource/eventsource
image

do we need to do this for it to function in IE11?

File: ui/packages/catalog-ui-search/src/main/webapp/js/EventSourceUtil.js (hunk 5 above)
Comment (cassandrabailey293) [reply ui/packages/catalog-ui-search/src/main/webapp/js/EventSourceUtil.js:14]: yes - refactoring this

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java (hunk 6 above)
Comment (cassandrabailey293) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java:62]: this is because we have no notification for when solr has updated the metacard. without this, the message will be sent too early, and when the client goes to fetch the new list of metacards, it won't yet be updated. 

the only way around this would be to implement solr notifications, something like this pr suggested(optimistic concurrency): https://github.com/codice/ddf/issues/4467

since all the metacard types i'm updating on are hard-committed, 3 seconds was my best guess at giving enough time.

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/sharing/container.tsx [hunk 11]
```diff
@@ -167,7 +171,7 @@ export class Sharing extends React.Component<Props, State> {
       e => e.value !== '' && e.category === Category.User
     )
 
-    if (this.state.isWorkspace && guest[0].access === 0) {
+    if (this.state.type === 'workspace' && guest[0].access === 0) {
```
Comment (rymach) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/sharing/container.tsx:174]: consider using an enum/constants for metacard types (both on the java side and js side)

**Comment by cassandrabailey293** [comment]
> does this reconnect if the connection fails?

Yes, you can also configure the time to wait before trying to reconnect: https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events. 

I'm currently trying to add some management code because there is a limit of 6 connections, so currently, if you build my pr and open 7 tabs , the 7th tab won't load properly. Part of this management code will likely be closing some connections as they aren't needed. But right now they will just always try to reconnect if disconnected AFAIK

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java (hunk 9 above)
Comment (cassandrabailey293) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java:48]: Yes :) changed it

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+            while (true) {
+              out.write(HEARTBEAT);
+              if (out.checkError()) {
+                // Subscriber error, break out of loop
+                break;
+              }
+              Thread.sleep(1000);
+            }
+            listeners.remove(out);
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java:47]: why do we synchronize `listeners.add` but not `listeners.remove`?  should the list itself be a synchronized list?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java [hunk 13]
```diff
@@ -226,6 +227,8 @@ public void init() {
         "/forms/:id",
         APPLICATION_JSON,
         (req, res) -> {
+          EventApplication.notifyListeners("searchform");
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java:230]: should we establish these as constants in the ui api module?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java (hunk 10 above)
Comment (cassandrabailey293) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java]: fixed

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/sharing/container.tsx (hunk 11 above)
Comment (cassandrabailey293) [reply ui/packages/catalog-ui-search/src/main/webapp/react-component/sharing/container.tsx]: done

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java (hunk 13 above)
Comment (cassandrabailey293) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java]: done

**Comment by cassandrabailey293** [comment]
> 1. how will this impact network perf?
> 2. will this be incredibly chatty if many changes to workspaces/resultforms/etc are happening?
> 3. all clients will now know when other users are changing workspaces/etc. it's a small detail, but it is something they wouldn't have known before

1. It shouldn't take a massive toll on network performance, since there are only a very small number of event sources allowed per browser anyway. For this initial implementation I'm basically just disabling the functionality when the 6th event source tries to get created - in other words, there should ever only be 5 connections at the most, per browser, from these event sources. So I guess it does depend on the number of simultaneous users/our load balancers/etc., **I wonder if there's a good way to stress test this somehow before it goes live? @bdeining ?** I've also set the heartbeat to send every 15 seconds, rather than every second. 
2. it could get chatty if a lot of changes are happening at once - though, now each type of event (result form, search form, workspace) triggers per-type, whereas originally i had them all firing all the time. perhaps this idea could be extended further to make the event type some combination of the event type and a unique session identifier, so each user only gets notifications for themselves. 
3. See above. We could "solve" this problem that way.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+            listeners.put(sourceId, out);
+            while (!out.checkError()) {
+              out.write(HEARTBEAT);
+              out.flush();
+              Thread.sleep(15000);
+            }
+            notifyListener(sourceId, EventType.CLOSE);
+            listeners.remove(sourceId);
+            return "Event Source connection closed";
+          } catch (Exception e) {
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java:66]: Do we have to catch a generic exception here, or can it be more specific?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+            if (listeners.size() >= 5) {
+              notifyListener(sourceId, EventType.CLOSE);
+              res.status(429);
+              return "Too many HTTP connections - cannot create new Event Source";
+            }
+            res.type("text/event-stream; charset=UTF-8");
+            res.header("Connection", "keep-alive");
+            res.header("Cache-Control", "no-cache");
+
+            PrintWriter out = res.raw().getWriter();
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java:55]: can res.raw() be null?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java (hunk 14 above)
Comment (cassandrabailey293) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java:66]: I'm not sure what kind of exception would be thrown here. I guess there would be some printwriter related exceptions but i'm not sure if that would be the only kind. I have yet to hit this catch block in my testing. Is it okay for me to leave it like this?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java (hunk 15 above)
Comment (cassandrabailey293) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/events/EventApplication.java:55]: i don't believe so - https://javadoc.io/static/com.sparkjava/spark-core/2.5.4/spark/Response.html#raw--

should i guard against it anyway?

**Comment by cassandrabailey293** [comment]
Closing due to https://github.com/codice/ddf-ui/pull/295#discussion_r463198297
